deployer/src/deployment/provisioner_factory.rs

```diff
@@ -85,82 +97,103 @@ pub enum ProvisionerError {
+impl<R: ResourceManager, S: SecretGetter> Factory for ProvisionerFactory<R, S> {
     async fn get_db_connection_string(
         &mut self,
         db_type: database::Type,
     ) -> Result<String, shuttle_service::Error> {
-        info!("Provisioning a {db_type} on the shuttle servers. This can take a while...");
-
         if let Some(ref info) = self.info {
             debug!("A database has already been provisioned for this deployment, so reusing it");
             return Ok(info.connection_string_private());
         }
 
         let r#type = ResourceType::Database(db_type.clone().into());
-        let db_type: DbType = db_type.into();
 
-        let request = Request::new(DatabaseRequest {
-            project_name: self.service_name.to_string(),
-            db_type: Some(db_type),
-        });
+        // Try to get the database info from provisioner if possible
+        let info = if let Some(claim) = self.claim.clone() {
+            info!("Provisioning a {db_type} on the shuttle servers. This can take a while...");
 
-        let response = self
-            .provisioner_client
-            .provision_database(request)
-            .await
-            .map_err(shuttle_service::error::CustomError::new)?
-            .into_inner();
+            let db_type: DbType = db_type.into();
 
-        let info: DatabaseReadyInfo = response.into();
-        let conn_str = info.connection_string_private();
+            let mut request = Request::new(DatabaseRequest {
+                project_name: self.service_name.to_string(),
+                db_type: Some(db_type),
+            });
+
+            request.extensions_mut().insert(claim);
 
-        self.resource_recorder
-            .insert_resource(&Resource {
-                service_id: self.service_id,
-                r#type,
-                data: serde_json::to_value(&info).unwrap(),
-            })
-            .await
-            .unwrap();
+            let response = self
+                .provisioner_client
+                .provision_database(request)
+                .await
+                .map_err(shuttle_service::error::CustomError::new)?
+                .into_inner();
+
+            let info: DatabaseReadyInfo = response.into();
+
+            self.resource_manager
+                .insert_resource(&Resource {
+                    service_id: self.service_id,
+                    r#type,
+                    data: serde_json::to_value(&info).map_err(|err| {
+                        shuttle_service::Error::Database(format!(
+                            "failed to convert DatabaseReadyInfo to json: {err}",
+                        ))
+                    })?,
+                })
+                .await
+                .map_err(|err| {
+                    shuttle_service::Error::Database(format!("failed to store resource: {err}"))
+                })?;
+
+            info
+        } else {
+            info!("Getting a {db_type} from a previous provision");
```

Comment:
I think this is okay, but in the future there may be a problem since resources will likely be scoped, if a users scope is reduced the old resource should not be restarted. If this used a refresh token, that problem may be easier to solve (unless I misunderstood the purpose of refresh tokens). However, we don't know when resources will be scoped, so I don't think we should worry about this yet.

Reply:
Yip, I agree